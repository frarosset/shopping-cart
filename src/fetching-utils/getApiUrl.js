const baseApiUrl = "https://dummyjson.com/products";

const queriesToString = (queryKeys, queries = {}) => {
  const sep = (first) => (first ? "?" : "&");

  const str = queryKeys.reduce((str, key) => {
    const val = queries[key];
    if (val != null) {
      str += `${sep(str === "")}${key}=${val}`;
    }
    return str;
  }, "");

  return str;
};

const getSingleProductApiUrl = (id, queries = {}) => {
  const queryKeys = ["select"];

  const queriesStr = queriesToString(queryKeys, queries);
  return encodeURI(`${baseApiUrl}/${id}${queriesStr}`);
};

const getAllProductsApiUrl = (queries = {}) => {
  const queryKeys = ["sortBy", "order", "limit", "skip", "select"];

  const queriesStr = queriesToString(queryKeys, queries);
  return encodeURI(`${baseApiUrl}${queriesStr}`);
};

const getCategoryProductsApiUrl = (category, queries = {}) => {
  const queryKeys = ["sortBy", "order", "limit", "skip", "select"];

  const queriesStr = queriesToString(queryKeys, queries);
  return encodeURI(`${baseApiUrl}/category/${category}${queriesStr}`);
};

const getSearchProductsApiUrl = (searchQuery, queries = {}) => {
  const queryKeys = ["q", "sortBy", "order", "limit", "skip", "select"];
  const queriesWithSearchQuery = Object.assign({}, { q: searchQuery }, queries);

  const queriesStr = queriesToString(queryKeys, queriesWithSearchQuery);
  return encodeURI(`${baseApiUrl}/search${queriesStr}`);
};

export {
  getSingleProductApiUrl,
  getAllProductsApiUrl,
  getCategoryProductsApiUrl,
  getSearchProductsApiUrl,
};
