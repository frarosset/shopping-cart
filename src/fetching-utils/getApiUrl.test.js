import { describe, it, expect } from "vitest";
import { getSingleProductApiUrl } from "./getApiUrl.js";

describe("getApiUrl", () => {
  describe("getSingleProductApiUrl", () => {
    const productId = 3;

    it(`returns the correct api url when an id (${productId}) is provided`, () => {
      const apiUrl = getSingleProductApiUrl(productId);
      expect(apiUrl).toMatchSnapshot();
    });
  });
});
