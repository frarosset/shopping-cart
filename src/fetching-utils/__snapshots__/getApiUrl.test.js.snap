// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getApiUrl > getAllProductsApiUrl > returns the correct api url 1`] = `"https://dummyjson.com/products"`;

exports[`getApiUrl > getAllProductsApiUrl > returns the correct api url when queries ({"limit":5,"skip":100,"select":"title,price","sortBy":"price","order":"asc"}) are provided 1`] = `"https://dummyjson.com/products?sortBy=price&order=asc&limit=5&skip=100&select=title,price"`;

exports[`getApiUrl > getSingleProductApiUrl > returns the correct api url when an id (3) and a select query ({"select":"title,price"}) are provided 1`] = `"https://dummyjson.com/products/3?select=title,price"`;

exports[`getApiUrl > getSingleProductApiUrl > returns the correct api url when an id (3) is provided 1`] = `"https://dummyjson.com/products/3"`;
